src/main/scala/com.snowplowanalytics.snowplow.analytics.scalasdk/Event.scala

```diff
@@ -0,0 +1,257 @@
+                 derived_tstamp:           Option[Instant],
+                 event_vendor:             Option[String],
+                 event_name:               Option[String],
+                 event_format:             Option[String],
+                 event_version:            Option[String],
+                 event_fingerprint:        Option[String],
+                 true_tstamp:              Option[Instant]) {
+
+  /**
+    * Extracts metadata from the event containing information about the types and Iglu URIs of its shred properties
+    */
+  def inventory: Set[Data.ShreddedType] = {
+    val unstructEvent = unstruct_event
+      .data
+      .toSet
+      .map((ue: SelfDescribingData[Json]) => Data.ShreddedType(Data.UnstructEvent, ue.schema))
+
+    val derivedContexts = derived_contexts
+      .data
+      .toSet
+      .map((ctx: SelfDescribingData[Json]) => Data.ShreddedType(Data.Contexts(Data.DerivedContexts), ctx.schema))
+
+    val customContexts = contexts
+      .data
+      .toSet
+      .map((ctx: SelfDescribingData[Json]) => Data.ShreddedType(Data.Contexts(Data.CustomContexts), ctx.schema))
+
+    customContexts ++ derivedContexts ++ unstructEvent
+  }
+
+  /**
+    * Returns the event as a map of keys to Circe JSON values
+    */
+  private def toJsonMap: Map[String, Json] = this.asJsonObject.toMap
+
+  /**
+    * Returns the event as a map of keys to Circe JSON values, while dropping inventory fields
+    */
+  def atomic: Map[String, Json] = toJsonMap - "contexts" - "unstruct_event" - "derived_contexts"
+
+  /**
+    * Returns the event as a list of key/Circe JSON value pairs.
+    * Unlike `toJsonMap` and `atomic`, these keys use the ordering of the canonical event model
+    */
+  def ordered: List[(String, Option[Json])] =
+    Event.parser.knownKeys.map(key => (key.name, toJsonMap.get(key.name)))
+
+  /**
+    * Returns a compound JSON field containing information about an event's latitude and longitude,
+    * or None if one of these fields doesn't exist
+    */
+  def geoLocation: Option[(String, Json)] =
+    for {
+      lat <- geo_latitude
+      lon <- geo_longitude
+    } yield "geo_location" -> s"$lat,$lon".asJson
+
+  /**
+    * Transforms the event to a validated JSON whose keys are the field names corresponding to the
+    * EnrichedEvent POJO of the Scala Common Enrich project. If the lossy argument is true, any
+    * self-describing events in the fields (unstruct_event, contexts and derived_contexts) are returned
+    * in a "shredded" format (e.g. "unstruct_event_com_acme_1_myField": "value"), otherwise a standard
+    * self-describing format is used.
+    *
+    * @param lossy Whether unstruct_event, contexts and derived_contexts should be flattened
+    */
+  def toJson(lossy: Boolean): Json =
+    if (lossy) {
+      JsonObject.fromMap(atomic ++ contexts.toJson.toMap ++ derived_contexts.toJson.toMap ++ unstruct_event.toJson.toMap ++ geoLocation).asJson
```

Comment:
Isn't `toJsonMap == atomic ++ contexts.toJson.toMap ++ derived_contexts.toJson.toMap ++ unstruct_event.toJson.toMap`?

Reply:
No - `toJson` isn't `asJson`, it's a method that converts contexts/unstruct. events into the lossy format.

Reply:
Gotcha! `asJson`/`toJson` is indeed slightly confusing.

Reply:
What about `toShreddedJson`?

Reply:
I like it. I don't care much about longer name as its rarely used API.